cdi: read XDG_CONFIG_HOME/cdi and XDG_RUNTIME_DIR/cdi for rootless

AkihiroSuda opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
#### Commit 1:  `cdi: read XDG_CONFIG_HOME/cdi and XDG_RUNTIME_DIR/cdi for rootless`

Fix #51601

#### Commit 2: `cdi: skip scanning non-readable dirs`

This simplifies `dockerd-rootless.sh` by removing the workaround for `CDI: Error associated with spec file /etc/cdi: failed to monitor for changes: permission denied`.

**- How I did it**

See the commits

**- How to verify it**

For commit 1 `cdi: read XDG_CONFIG_HOME/cdi and XDG_RUNTIME_DIR/cdi for rootless`, create `~/.config/cdi/foo.yaml` as follows, and run `docker run --rm --device example.com/foo=foo alpine ls -l /dev/null2`

```yaml
cdiVersion: "0.6.0"
kind: "example.com/foo"
devices:
- name: foo
  containerEdits:
    mounts:
    - hostPath: /dev/null
      containerPath: /dev/null2
      options: [bind]
```

Try the same with `/run/user/$UID/cdi`, `/etc/cdi`, and `/var/run/cdi` too.

For commit 2 `cdi: skip scanning non-readable dirs`:

```bash
sudo mkdir -p -m /etc/cdi
dockerd-rootless-setuptool.sh install
docker info
# no error
```

**- Human readable description for the release notes**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section.

NOTE: Only fill this section if changes introduced in this PR are user-facing.
The PR must have a relevant impact/ label.
-->
```markdown changelog
Rootless: Consider `$XDG_CONFIG_HOME/cdi` and `$XDG_RUNTIME_DIR/cdi` when looking for CDI devices
```

**- A picture of a cute animal (not mandatory but encouraged)**

AkihiroSuda committed:
cdi: read XDG_CONFIG_HOME/cdi and XDG_RUNTIME_DIR/cdi for rootless

Fix issue 51601

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda committed:
cdi: skip scanning non-readable dirs

This simplifies `dockerd-rootless.sh` by removing the workaround for
`CDI: Error associated with spec file /etc/cdi: failed to monitor for changes: permission denied`.

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

vvoland commented on contrib/dockerd-rootless.sh:
```diff
@@ -190,24 +190,6 @@ if [ -z "$_DOCKERD_ROOTLESS_CHILD" ]; then
-	# mode; otherwise the Docker daemon won't have access to the CDI
-	# specification files.
-	mount_directory /etc/cdi
-	mount_directory /var/run/cdi
```
Wasn't this required to make the host definition available to the rootless daemon?

AkihiroSuda committed:
cdi: resolve symlinks

Fix using /etc/cdi entries in rootless mode:

> level=warning msg="Refreshing the CDI registry generated errors"
> error="failed to monitor for changes: no such file or directory\nfailed to monitor for changes: no such file or directory"

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda committed:
integration: cdi: add TestEtcCDI to verify /etc/cdi is used in rootless mode

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

vvoland reviewed (approved):
LGTM, thanks!

thaJeztah reviewed (approved):
LGTM
